msgpack/_unpacker.pyx

```diff
@@ -129,12 +131,27 @@ def unpackb(object packed, object object_hook=None, object list_hook=None,
     cdef Py_ssize_t off = 0
     cdef int ret
 
-    cdef char* buf
+    cdef Py_buffer view
+    cdef char* buf = NULL
     cdef Py_ssize_t buf_len
     cdef char* cenc = NULL
     cdef char* cerr = NULL
-
-    PyObject_AsReadBuffer(packed, <const void**>&buf, &buf_len)
+    cdef char buffer_supported = 0
+
+    if PyObject_CheckBuffer(packed):
+        buffer_supported = 1
+        if PyObject_GetBuffer(packed, &view, PyBUF_SIMPLE) == 0:
+            if view.itemsize != 1:
+                PyBuffer_Release(&view)
+                raise ValueError("cannot unpack from multi-byte object")
+            buf_len = view.len
+            buf = <char*> view.buf
+    else:
+        PyObject_AsReadBuffer(packed, <const void**>&buf, &buf_len)
+        PyErr_WarnEx(DeprecationWarning,
+                     "unpacking %s requires old buffer protocol,"
+                     "which will be removed in msgpack 1.0" % type(packed),
```

Comment:
Is there any reference about convention of exception message?
```
>>> "3" + 2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't convert 'int' object to str implicitly
```

in `_iomodule.c` file:

```
    /* Parameters validation */
    if (universal) {
        if (creating || writing || appending || updating) {
            PyErr_SetString(PyExc_ValueError,
                            "mode U cannot be combined with x', 'w', 'a', or '+'");
            return NULL;
        }
        if (PyErr_WarnEx(PyExc_DeprecationWarning,
                         "'U' mode is deprecated", 1) < 0)
            return NULL;
        reading = 1;
    }

    if (text && binary) {
        PyErr_SetString(PyExc_ValueError,
                        "can't have text and binary mode at once");
        return NULL;
    }

    if (creating + reading + writing + appending > 1) {
        PyErr_SetString(PyExc_ValueError,
                        "must have exactly one of create/read/write/append mode");
        return NULL;
    }

    if (binary && encoding != NULL) {
        PyErr_SetString(PyExc_ValueError,
                        "binary mode doesn't take an encoding argument");
        return NULL;
    }

    if (binary && errors != NULL) {
        PyErr_SetString(PyExc_ValueError,
                        "binary mode doesn't take an errors argument");
```

So I feel last dot is omitted by convention, at least one line message.